### DISCUSSION TITLE
Using dyn4j with MATLAB for scientific research

### DISCUSSION THREAD

**Discussion by iandol** [dyn4j/dyn4j#297] [Q&A] [answered]
Hi, first off I apologise for what may be very basic questions. I am not a Java programmer but use MATLAB. I am doing research on a question many of you may find interesting: do biological brains have a built-in "intuitive physics" engine? MATLAB runs the Psychophysics toolbox, a widely used framework in which to run precise scientific experiments, hence why we an others use MATLAB. After having scaling issues with a simple engine I made myself, I found dyn4j. MATLAB can call Java like so:

```matlab
thisShape = javaObject('org.dyn4j.geometry.Rectangle', bw, bh);
```

I have a working system that can add bodies to a world and run a simulation and dyn4j is brilliant so thank you! There are some issues I am having problems with:

1) I need to get collison information. The samples (e.g. basketball one) seem to use java callbacks as best practice and I don't understand how to implement these; the java syntax itself is not clear as I don't know it or how it can be used with MATLAB. Is there some simpler way to do this, just get some list of collisions on each step exported by a java method to a list/array. There are lots of named methods in the world class and I am not clear what is best to use.
2) I need to rewind time: change a start-condition parameter then run the world again. I tried to use `update(0.0)` or find an accumulated time or total steps property that could be used, but didn't find the best way to do this. I need a dyn4j time machine 😄 

Thank you for a great physics engine!

**Comment by iandol** [comment]
To partially answer myself for (1) -- `world.getContacts(body)` gets me a non-empty list if that `body` has collided with another at a particular step in the simulation. A naive question for this, if an object has high velocity, at step X it may be just before a contact and at step X+1 it may be just after a contact, i.e. the step resolution can miss the collision? What happens in this case? My step resolution is 16ms as that is my display framerate.

**Comment by wnbittle** [comment, marked as answer]
@iandol  Good questions and appreciate the feedback!

1. `world.getContacts(body)` is a good alternative to the callback system.  Hopefully that gives you what you are looking for.  There's also `world.getCollisionDataIterator()` that might be useful - here's how it would be used in Java:

```java
Iterator<WorldCollisionData> it = world.getCollisionDataIterator();
while (it.hasNext()) {
  WorldCollisionData data = it.next();
  // do something with data
}
```

The idea of the callback system is to create an object like a CollisionListener or ContactListener and implement the methods that you are interested in.  In the those methods (or events), you would save off the information that you care about in another place of your own design.  Sadly I won't be much use when it comes to MATLAB.  I'm familiar with the product, but have very little experience with it (more than 10 years ago).

To your question about the discrete nature of the engine and missed collisions - this is an unfortunate side effect of most physics engines today.  The general solution to the problem is something called Continuous Collision Detection (CCD).  dyn4j has a rather weak implementation of this to stay performant.  It should be enabled by default for infinite mass objects against non-infinite mass objects.  You can also flag a body as a "bullet" to purposefully say "Do CCD on this all the time".  Now, that said, I say the implementation is weak because it uses a time-stealing method.  The behavior is a check after discrete collision detection to see if we missed something, if we did, we roll back to the time of impact (that was missed).  The effect is a strange pausing when it's detected and resolved.  There's been some research work done in this area (not by me) and a few new ways to handle this without the pausing/time-stealing.

2. Unfortunately, there isn't a rewind capability.  Typically, what's done is the initial state of the World would be put into a function (which is what the samples do), for example:

```java
public void setup(World world) {
  // create bodies
  // create joints
  // do all the setup
  // etc.
}
```

Then, just create a new world every time you need to restart:
```java
World w = new World();
setup(w);
// ... run the simulation

World w2 = new World();
setup(w2);
// ... run the simulation
```

If you need to run for a bit, then save and reload that state, that's a lot harder.  There's no dyn4j native capability for that at the moment.  I've been looking into it recently, but ran into a few design challenges.  For this, I had to manually write the code to save/read what I cared about to a file.

**Comment by iandol** [comment]
@wnbittle -- thank you so much for your detailed and helpful replies!

Regarding collisions, as my requirements are minimal, and I don't think I can run callbacks in MATLAB, the `getContacts` method is the best solution. I really appreciate that you have functions for this in addition to the callbacks (which I understand are a better/more elegant way to do this). One question I do have is what happens when a body has multiple collisions, in some preliminary testing I only ever see body1 and body2 in the returned structure?

And regarding time-rewinding, I have rewritten my MATLAB wrapper to rebuild new worlds for each new trial of my experiment and it works well.

**Comment by wnbittle** [comment]
I'm glad to hear the library is useful and you've been able to get things to work!  The `getContacts(Body)` method should return all the contacts for the given body.  If body1 is interacting with Body2 and Body3, then you should get two `ContactConstraint`s, one with Body1 and Body2 and another with Body1 and Body3.
